Further reduce use of memset()

cdumez opened the pull request:
#### 52af84aa93ee45a39f778c9006f78005d54afb94
<pre>
Further reduce use of memset()
<a href="https://bugs.webkit.org/show_bug.cgi?id=284684">https://bugs.webkit.org/show_bug.cgi?id=284684</a>

Reviewed by Darin Adler.

Further reduce use of memset(), in favor of safer alternatives.

* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.h:
(JSC::JSGenericTypedArrayView::typedSpan const):
(JSC::JSGenericTypedArrayView::typedSpan):
* Source/WTF/wtf/GregorianDateTime.h:
* Source/WTF/wtf/HashTable.h:
(WTF::HashTableBucketInitializer&lt;true&gt;::initialize):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::memcpySpan):
(WTF::memmoveSpan):
(WTF::zeroSpan):
(WTF::zeroBytes):
* Source/WebCore/Modules/webaudio/AudioBufferSourceNode.cpp:
(WebCore::AudioBufferSourceNode::process):
(WebCore::AudioBufferSourceNode::renderSilenceAndFinishIfNotLooping):
(WebCore::AudioBufferSourceNode::renderFromBuffer):
(WebCore::AudioBufferSourceNode::setBufferForBindings):
* Source/WebCore/Modules/webaudio/AudioBufferSourceNode.h:
* Source/WebCore/Modules/webaudio/AudioWorkletProcessor.cpp:
(WebCore::zeroJSArray):
* Source/WebCore/Modules/webaudio/MediaStreamAudioSourceCocoa.cpp:
(WebCore::copyChannelData):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
(WebCore::AudioSampleBufferList::zeroABL):
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
(WebCore::ZeroABL):
* Source/WebCore/platform/audio/cocoa/SpanCoreAudio.h: Added.
(WebCore::mDataMutableByteSpan):
(WebCore::mDataByteSpan):
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.mm:
(WebCore::AudioSourceProviderAVFObjC::process):
* Source/WebKit/Shared/Cocoa/AuxiliaryProcessCocoa.mm:
(WebKit::AuxiliaryProcess::registerWithStateDumper):

Canonical link: <a href="https://commits.webkit.org/287848@main">https://commits.webkit.org/287848@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c2dd6584074519cfbfae7f9387c420cb803b9950

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81097 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/622 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35040 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85626 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32083 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83207 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/640 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8429 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63310 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21076 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84166 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/403 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73816 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43608 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/300 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27978 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30541 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74076 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71829 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28535 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87061 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80155 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8327 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5897 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71616 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8504 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69651 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70851 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14898 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13824 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/102561 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12571 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8288 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13811 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24921 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-no-cjit (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8125 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11645 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9933 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/Modules/webaudio/AudioBufferSourceNode.h:
```diff
@@ -103,8 +103,8 @@ class AudioBufferSourceNode final : public AudioScheduledSourceNode {
     RefPtr<AudioBuffer> m_buffer WTF_GUARDED_BY_LOCK(m_processLock); // Only modified on the main thread but used on the audio thread.
 
     // Pointers for the buffer and destination.
```
The word "pointers" in this comment isn't great.

darinadler commented on Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
```diff
@@ -146,7 +147,10 @@ inline void ZeroABL(AudioBufferList* list, size_t destOffset, size_t nbytes)
         if (destOffset > dest->mDataByteSize)
             continue;
-        memset(static_cast<Byte*>(dest->mData) + destOffset, 0, std::min<size_t>(nbytes, dest->mDataByteSize - destOffset));
+        auto mDataSpan = mDataMutableByteSpan(*dest).subspan(destOffset);
```
I don't think we should have "m" in the name of this local variable. I think the "m" is for data member.

darinadler commented on Source/WebCore/platform/audio/cocoa/SpanCoreAudio.h:
```diff
@@ -0,0 +1,44 @@
+
+namespace WebCore {
+
+inline std::span<uint8_t> mDataMutableByteSpan(AudioBuffer& buffer)
```
I don't think we should have "m" in the name of this function. I think the "m" is for data member.

darinadler commented on Source/WebCore/platform/audio/cocoa/SpanCoreAudio.h:
```diff
@@ -0,0 +1,44 @@
+    return unsafeMakeSpan(static_cast<uint8_t*>(buffer.mData), buffer.mDataByteSize);
+}
+
+inline std::span<const uint8_t> mDataByteSpan(const AudioBuffer& buffer)
```
I don't think we should have "m" in the name of this function. I think the "m" is for data member.

cdumez committed:
Further reduce use of memset()
https://bugs.webkit.org/show_bug.cgi?id=284684

Reviewed by Darin Adler.

Further reduce use of memset(), in favor of safer alternatives.

* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.h:
(JSC::JSGenericTypedArrayView::typedSpan const):
(JSC::JSGenericTypedArrayView::typedSpan):
* Source/WTF/wtf/GregorianDateTime.h:
* Source/WTF/wtf/HashTable.h:
(WTF::HashTableBucketInitializer<true>::initialize):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::memcpySpan):
(WTF::memmoveSpan):
(WTF::zeroSpan):
(WTF::zeroBytes):
* Source/WebCore/Modules/webaudio/AudioBufferSourceNode.cpp:
(WebCore::AudioBufferSourceNode::process):
(WebCore::AudioBufferSourceNode::renderSilenceAndFinishIfNotLooping):
(WebCore::AudioBufferSourceNode::renderFromBuffer):
(WebCore::AudioBufferSourceNode::setBufferForBindings):
* Source/WebCore/Modules/webaudio/AudioBufferSourceNode.h:
* Source/WebCore/Modules/webaudio/AudioWorkletProcessor.cpp:
(WebCore::zeroJSArray):
* Source/WebCore/Modules/webaudio/MediaStreamAudioSourceCocoa.cpp:
(WebCore::copyChannelData):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
(WebCore::AudioSampleBufferList::zeroABL):
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
(WebCore::ZeroABL):
* Source/WebCore/platform/audio/cocoa/SpanCoreAudio.h: Added.
(WebCore::mDataMutableByteSpan):
(WebCore::mDataByteSpan):
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.mm:
(WebCore::AudioSourceProviderAVFObjC::process):
* Source/WebKit/Shared/Cocoa/AuxiliaryProcessCocoa.mm:
(WebKit::AuxiliaryProcess::registerWithStateDumper):

Canonical link: https://commits.webkit.org/287848@main
